tools/swift-inspect/Sources/SwiftInspectLinux/ElfFile.swift

```diff
@@ -0,0 +1,184 @@
+
+    let ehdrSize = isElf64 ? Elf64_Ehdr.symbolSize : Elf32_Ehdr.symbolSize
+    file.seek(toFileOffset: 0)
+    guard let ehdrData = try file.read(upToCount: ehdrSize), ehdrData.count == ehdrSize else {
+      file.closeFile()
+      throw Error.FileReadFailure(filePath, offset: 0, size: UInt64(ehdrSize))
+    }
+
+    if isElf64 {
+      self.ehdr = ehdrData.withUnsafeBytes { $0.load(as: Elf64_Ehdr.self) as ElfEhdr }
+    } else {
+      self.ehdr = ehdrData.withUnsafeBytes { $0.load(as: Elf32_Ehdr.self) as ElfEhdr }
+    }
+  }
+
+  deinit { file.closeFile() }
+
+  // returns a map of symbol names to their offset range in file (+ baseAddress)
+  public func loadSymbols(baseAddress: UInt64 = 0) throws -> SymbolMap {
+    guard let sectionCount = UInt(exactly: self.ehdr.shnum) else {
+      throw Error.MalformedElfFile(
+        self.filePath, description: "invalid ehdr.shnum: \(self.ehdr.shnum)")
+    }
+
+    var symbols: SymbolMap = [:]
+    for sectionIndex in 0..<sectionCount {
+      let shdr: ElfShdr =
+        isElf64
+        ? try self.readShdr(index: sectionIndex) as Elf64_Shdr
+        : try self.readShdr(index: sectionIndex) as Elf32_Shdr
+
+      guard shdr.type == SHT_SYMTAB || shdr.type == SHT_DYNSYM else { continue }
+
+      let sectionData: Data = try self.readSection(shdr: shdr)
+      let symTable: [ElfSym] =
+        self.isElf64
+        ? sectionData.withUnsafeBytes { Array($0.bindMemory(to: Elf64_Sym.self)) }
+        : sectionData.withUnsafeBytes { Array($0.bindMemory(to: Elf32_Sym.self)) }
+
+      guard shdr.entsize == (self.isElf64 ? Elf64_Sym.symbolSize : Elf32_Sym.symbolSize) else {
+        throw Error.MalformedElfFile(self.filePath, description: "invalid ElfShdr.sh_entsize")
+      }
+
+      // the link field in the section header for a symbol table section refers
+      // to the index of the string table section containing the symbol names
+      guard let linkIndex = UInt(exactly: shdr.link) else {
+        throw Error.MalformedElfFile(
+            self.filePath, description: "invalid ElfShdr.sh_link: \(shdr.link)")
+      }
+
+      let shdrLink: ElfShdr =
+        isElf64
+        ? try self.readShdr(index: UInt(linkIndex)) as Elf64_Shdr
+        : try self.readShdr(index: UInt(linkIndex)) as Elf32_Shdr
+
+      guard shdrLink.type == SHT_STRTAB else {
+        throw Error.MalformedElfFile(self.filePath, description: "linked section not SHT_STRTAB")
+      }
+
+      // load the entire contents of the string table into memory
+      let strTable: Data = try self.readSection(shdr: shdrLink)
+
+      let symCount = Int(shdr.size / shdr.entsize)
+      for symIndex in 0..<symCount {
+        let sym = symTable[symIndex]
+        guard sym.shndx != SHN_UNDEF, sym.value != 0, sym.size != 0 else { continue }
+
+        // sym.name is a byte offset into the string table
+        guard let strStart = Int(exactly: sym.name), strStart < strTable.count else {
+          throw Error.MalformedElfFile(
```

Comment:
Is this safe? There is no guarantee that there is a string table in the binary.

Reply:
This bit is safe because it won't even get here if there's no string table to load, but its a fair point that an Elf file may not have symbol or string table sections.

Reply:
fwiw, I implemented the `loadSymbolTable` function to return an empty symbol table if it doesn't find a symtab or dynsym section to load symbols from